### ISSUE TITLE
Erroneous reportUnhashable on child of class inheriting Pydantic BaseModel, even with explicit __eq__ and __hash__

### ISSUE THREAD

**Issue by DylanLukes** [microsoft/pyright#9249]
**Description**

The presence of `BaseModel` as a superclass leads Pyright to erroneously claim that `A()` below is unhashable, whereas `B()` is correctly identified as hashable. In both cases **the `__eq__` and `__hash__` dunders have been explicitly defined in the base class.

Moreover, Pyright does not complain about adding the `Hashable` protocol to `A`. It appears that this issue occurs inconsistently in different sites.

As for _why_ this might be happening with `BaseModel`, I'm not sure. My first suspicion was that it has to do with how Pydantic abuses `if [not] TYPE_CHECKING`, but trying to minimize `BaseModel` to my own little `MyBaseModel` didn't  prove fruitful.

I also can't reproduce the issue with mypy 1.12.0 (`--strict`).

**Code**

```python
from pydantic import BaseModel


class BaseA(BaseModel):
    def __eq__(self, other: object) -> bool:
        return isinstance(other, self.__class__)

    def __hash__(self) -> int:
        return hash(self.__class__)


class A(BaseA, Hashable):  # ok: no type checker complaint about Hashable
    pass

class BaseB:
    def __eq__(self, other: object) -> bool:
        return isinstance(other, self.__class__)

    def __hash__(self) -> int:
        return hash(self.__class__)

class B(BaseB):
    pass

_1 = {BaseA()}  # OK
_2 = {BaseB()}  # OK
_3 = {A()}  # Pyright: Set entry must be hashable. Type "A" is not hashable (reportUnhashable)
_4 = {B()}  # OK
```

**Third-Party Dependencies and System**

```sh
❯ python -c "import pydantic.version; print(pydantic.version.version_info())"
             pydantic version: 2.9.2
        pydantic-core version: 2.23.4
          pydantic-core build: profile=release pgo=false
                 install path: /Users/dylan/PycharmProjects/renkon/.hatch/renkon/lib/python3.13/site-packages/pydantic
               python version: 3.13.0 (main, Oct  7 2024, 05:02:14) [Clang 16.0.0 (clang-1600.0.26.3)]
                     platform: macOS-15.1-arm64-arm-64bit-Mach-O
             related packages: pydantic-settings-2.6.0 fastapi-0.115.2 mypy-1.12.0 typing_extensions-4.12.2
                       commit: unknown
```

**Comment by erictraut** [comment]
Hashability is problematic when it comes to static typing. There are a number of things going on here.

Let's first consider the `Hashable` protocol class exported from the `typing` module. A class that is compatible with the `Hashable` protocol simply needs to define a method named `__hash__` or derive from a class that defines such a method. In your code sample, `A` satisfies this protocol, so there is no type error generated when `A` is declared to derive from `Hashable`.

The runtime has a different notion of hashability. You can read about these rules [in this documentation](https://docs.python.org/3/reference/datamodel.html#object.__hash__). The most relevant part of this documentation is:

> A class that overrides `__eq__` and does not define `__hash__` will have its `__hash__` implicitly set to None.

This means `__hash__` (and therefore hashability) doesn't follow the normal rules of inheritance — the rules that the Python type system is based on.

Pyright attempts to model these runtime rules when determining whether a type is "hashable" (lower-case). This is different from whether a type conforms with the `Hashable` (upper-case) protocol.

So, why does your code sample run without an exception at runtime even though pyright indicates that it's not "hashable"? That's because pydantic's `BaseModel` is deviating from the way that dataclass works in a subtle way. Why does that matter? Because pydantic uses a feature of the type system called `dataclass_transform` to tell static standards-compliant type checkers like pyright what behaviors they should assume. You can read more about it [here](https://typing.readthedocs.io/en/latest/spec/dataclasses.html#the-dataclass-transform-decorator). The spec says that if a library uses `dataclass_transform`, type checkers should assume that it applies dataclass semantics. Unfortunately, pydantic is deviating from those semantics in a subtle way.

Here's a slightly modified version of your code sample that demonstrates the issue.

```python
from typing import Hashable
from pydantic import BaseModel

class BaseA(BaseModel):
    def __eq__(self, other: object) -> bool:
        return isinstance(other, self.__class__)

    def __hash__(self) -> int:
        return hash(self.__class__)

class A(BaseA, Hashable):
    x: int

class BaseB:
    def __eq__(self, other: object) -> bool:
        return isinstance(other, self.__class__)

    def __hash__(self) -> int:
        return hash(self.__class__)

class B(BaseB):
    pass

_1 = {BaseA()}  # OK
_2 = {BaseB()}  # OK
_3 = {A(x=1)}
_4 = {B()}
```

Pydantic is implemented with a metaclass (`ModelMetaclass`, which is the metaclass for `BaseClass`) rather than a decorator, which is used for stdlib `dataclass`, but the resulting behaviors should be the same. Here's the comparable code if we were to use a dataclass.

```python
from dataclasses import dataclass
from typing import Hashable

@dataclass
class BaseA:
    def __eq__(self, other: object) -> bool:
        return isinstance(other, self.__class__)

    def __hash__(self) -> int:
        return hash(self.__class__)

@dataclass
class A(BaseA, Hashable):
    x: int

class BaseB:
    def __eq__(self, other: object) -> bool:
        return isinstance(other, self.__class__)

    def __hash__(self) -> int:
        return hash(self.__class__)

class B(BaseB):
    pass

_1 = {BaseA()}
_2 = {BaseB()}
_3 = {A(x=1)}
_4 = {B()}
```

If you run the dataclass code, you'll see that it crashes with a `TypeError: unhashable type: 'A'` exception.

It appears that pydantic's runtime behavior deviates from `dataclass` in that it synthesizes a `__hash__` method for class `A`. A static standards-compliant type checker like pyright would have no way of know about this non-standard behavior.

I hope that makes sense.

I'm going to close this issue because I think pyright is working correctly here. If you'd like to take this up with the maintainers of pydantic, you're welcome to do so.

**Comment by DylanLukes** [comment]
﻿
Wow, thank you so much for the detailed reply. I'll need to take some time to unpack it. I did originally intend to take this up as a Pydantic issue, I'll give that shot.

As a temporary workaround I can add delegating dunders that just call the super and Pyright is content with that. 

> It appears that pydantic's runtime behavior deviates from dataclass in that it synthesizes a __hash__ method for class A.

My understanding is that this is _not_ the case unless `frozen = true` is specified in the configuration (nested `Config` or class kwargs).

— Dylan
